File: rtls-kmp/rtls-websocket/src/commonMain/kotlin/com/rtls/websocket/RealTimeLocationClient.kt
package com.rtls.websocket

import com.rtls.core.LocationPoint
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID
import kotlin.math.min
import kotlin.math.pow

sealed class RealTimeEvent {
    data class Connected(val url: String) : RealTimeEvent()
    data object Disconnected : RealTimeEvent()
    data class Reconnecting(val attempt: Int, val delayMs: Long) : RealTimeEvent()
    data class LocationReceived(val point: LocationPoint) : RealTimeEvent()
    data class PushAcknowledged(val reqId: String, val status: String) : RealTimeEvent()
    data class BatchAcknowledged(val reqId: String, val acceptedIds: List<String>) : RealTimeEvent()
    data class Error(val message: String) : RealTimeEvent()
    data class Subscribed(val userId: String) : RealTimeEvent()
}

/**
 * High-level real-time location client over WebSocket.
 * Independently usable — no offline sync or local storage required.
 *
 * Features:
 * - Push locations in real-time (fire-and-forget or ack-based)
 * - Subscribe to other users' location updates
 * - Auto-reconnect with exponential backoff
 * - Heartbeat ping/pong
 */
class RealTimeLocationClient(
    private val config: WebSocketConfig,
    private val channel: RealTimeChannel,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
) {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    private val _events = MutableSharedFlow<RealTimeEvent>(extraBufferCapacity = 256)
    val events: SharedFlow<RealTimeEvent> = _events.asSharedFlow()

    private val _incomingLocations = MutableSharedFlow<LocationPoint>(extraBufferCapacity = 256)
    val incomingLocations: SharedFlow<LocationPoint> = _incomingLocations.asSharedFlow()

    private var messageJob: Job? = null
    private var pingJob: Job? = null
    private var reconnectJob: Job? = null
    private var running = false
    private var isBackgroundMode = false
    private var reconnectAttempt = 0
    private val subscribedUserIds = mutableSetOf<String>()

    suspend fun connect() {
        running = true
        reconnectAttempt = 0
        doConnect()
    }

    suspend fun disconnect() {
        running = false
        reconnectJob?.cancel()
        pingJob?.cancel()
        messageJob?.cancel()
        try { channel.disconnect() } catch (_: Exception) {}
        _events.emit(RealTimeEvent.Disconnected)
    }

    suspend fun pushLocation(point: LocationPoint) {
        val msg = WsLocationPush(reqId = UUID.randomUUID().toString(), point = point)
        channel.send(json.encodeToString(msg))
    }

    suspend fun pushBatch(points: List<LocationPoint>) {
        val msg = WsLocationBatch(reqId = UUID.randomUUID().toString(), points = points)
        channel.send(json.encodeToString(msg))
    }

    suspend fun subscribe(userId: String) {
        subscribedUserIds.add(userId)
        if (channel.isConnected) {
            channel.send(json.encodeToString(WsSubscribe(userId = userId)))
        }
    }

    suspend fun unsubscribe(userId: String) {
        subscribedUserIds.remove(userId)
        if (channel.isConnected) {
            channel.send(json.encodeToString(WsUnsubscribe(userId = userId)))
        }
    }

    /**
     * Switch between foreground/background ping intervals.
     * In background mode, ping interval increases to reduce radio wake-ups.
     * If [WebSocketConfig.disconnectInBackground] is enabled, the WebSocket is fully disconnected.
     */
    suspend fun setBackgroundMode(enabled: Boolean) {
        if (isBackgroundMode == enabled) return
        isBackgroundMode = enabled

        if (enabled && config.disconnectInBackground) {
            disconnect()
            return
        }
        if (!enabled && config.disconnectInBackground && !running) {
            connect()
            return
        }
        // Restart ping loop with appropriate interval
        if (running && channel.isConnected) {
            startPingLoop()
        }
    }

    private suspend fun doConnect() {
        try {
            val wsUrl = buildWsUrl(config.baseUrl)
            val token = config.tokenProvider.accessToken()
            val headers = mutableMapOf<String, String>()
            if (token.isNotBlank()) headers["Authorization"] = "Bearer $token"

            channel.connect(wsUrl, headers)

            val authMsg = WsAuthMessage(token = token)
            channel.send(json.encodeToString(authMsg))

            reconnectAttempt = 0
            _events.emit(RealTimeEvent.Connected(wsUrl))

            for (uid in subscribedUserIds) {
                channel.send(json.encodeToString(WsSubscribe(userId = uid)))
            }

            startMessageLoop()
            startPingLoop()
        } catch (e: Exception) {
            _events.emit(RealTimeEvent.Error(e.message ?: "Connection failed"))
            if (running && config.autoReconnect) scheduleReconnect()
        }
    }

    private fun startMessageLoop() {
        messageJob?.cancel()
        messageJob = scope.launch {
            try {
                channel.incomingMessages.collect { raw ->
                    handleServerMessage(raw)
                }
            } catch (e: Exception) {
                if (running) {
                    _events.emit(RealTimeEvent.Error(e.message ?: "Connection lost"))
                    _events.emit(RealTimeEvent.Disconnected)
                    if (config.autoReconnect) scheduleReconnect()
                }
            }
        }
    }

    private fun startPingLoop() {
        pingJob?.cancel()
        pingJob = scope.launch {
            while (isActive && channel.isConnected) {
                val interval = activePingIntervalMs
                delay(interval)
                try { channel.send(json.encodeToString(WsPing())) } catch (_: Exception) { break }
            }
        }
    }

    private val activePingIntervalMs: Long
        get() = if (!isBackgroundMode) {
            config.pingIntervalMs
        } else {
            // In background mode, use cellular interval by default for maximum battery savings
            // For full cellular-aware logic, pass NetworkMonitor and check connection type
            config.cellularBackgroundPingIntervalMs
        }

    private suspend fun handleServerMessage(raw: String) {
        try {
            val msg = json.decodeFromString<WsServerMessage>(raw)
            when (msg.type) {
                "auth.ok" -> {} // authenticated
                "location.ack" -> {
                    _events.emit(RealTimeEvent.PushAcknowledged(msg.reqId ?: "", msg.status ?: "accepted"))
                }
                "location.batch_ack" -> {
                    _events.emit(RealTimeEvent.BatchAcknowledged(msg.reqId ?: "", msg.acceptedIds ?: emptyList()))
                }
                "location.update", "location" -> {
                    msg.point?.let { point ->
                        _incomingLocations.emit(point)
                        _events.emit(RealTimeEvent.LocationReceived(point))
                    }
                }
                "subscribed" -> {
                    _events.emit(RealTimeEvent.Subscribed(msg.userId ?: ""))
                }
                "pong" -> {} // heartbeat response
                "error" -> {
                    _events.emit(RealTimeEvent.Error(msg.message ?: "Server error"))
                }
            }
        } catch (e: Exception) {
            _events.emit(RealTimeEvent.Error("Failed to parse server message: ${e.message}"))
        }
    }

    private fun scheduleReconnect() {
        reconnectJob?.cancel()
        reconnectJob = scope.launch {
            reconnectAttempt++
            val exp = 2.0.pow((reconnectAttempt - 1).coerceAtMost(10).toDouble())
            val delayMs = min(config.reconnectMaxDelayMs, (config.reconnectBaseDelayMs * exp).toLong())
            _events.emit(RealTimeEvent.Reconnecting(reconnectAttempt, delayMs))
            delay(delayMs)
            if (running) doConnect()
        }
    }

    private fun buildWsUrl(baseUrl: String): String {
        val trimmed = baseUrl.trimEnd('/')
        val withPath = if (trimmed.endsWith("/v1/ws")) trimmed else "$trimmed/v1/ws"
        return withPath
            .replace("https://", "wss://")
            .replace("http://", "ws://")
            .let { if (!it.startsWith("ws")) "ws://$it" else it }
    }
}

File: Sources/RTLSWebSocket/RealTimeLocationClient.swift
import Foundation
import RTLSCore

/// High-level real-time location client over WebSocket.
/// Independently usable — no offline sync or local storage required.
///
/// Features:
/// - Push locations in real-time
/// - Subscribe to other users' location updates
/// - Auto-reconnect with exponential backoff
/// - Heartbeat ping/pong
public actor RealTimeLocationClient {
    public struct Configuration: Sendable {
        public var baseURL: URL
        public var tokenProvider: AuthTokenProvider
        public var autoReconnect: Bool
        public var reconnectBaseDelay: TimeInterval
        public var reconnectMaxDelay: TimeInterval
        public var pingInterval: TimeInterval
        /// Ping interval used in background mode. Longer intervals allow cellular radio dormancy.
        public var backgroundPingInterval: TimeInterval
        /// When true, fully disconnect WebSocket when entering background.
        public var disconnectInBackground: Bool

        public init(
            baseURL: URL,
            tokenProvider: AuthTokenProvider,
            autoReconnect: Bool = true,
            reconnectBaseDelay: TimeInterval = 1.0,
            reconnectMaxDelay: TimeInterval = 30.0,
            pingInterval: TimeInterval = 30.0,
            backgroundPingInterval: TimeInterval = 120.0,
            disconnectInBackground: Bool = false
        ) {
            self.baseURL = baseURL
            self.tokenProvider = tokenProvider
            self.autoReconnect = autoReconnect
            self.reconnectBaseDelay = reconnectBaseDelay
            self.reconnectMaxDelay = reconnectMaxDelay
            self.pingInterval = pingInterval
            self.backgroundPingInterval = backgroundPingInterval
            self.disconnectInBackground = disconnectInBackground
        }
    }

    public enum Event: Sendable {
        case connected
        case disconnected
        case reconnecting(attempt: Int, delay: TimeInterval)
        case locationReceived(LocationPoint)
        case pushAcknowledged(reqId: String, status: String)
        case batchAcknowledged(reqId: String, acceptedIds: [String])
        case subscribed(userId: String)
        case error(String)
    }

    private let config: Configuration
    private let session: URLSession

    private var webSocketTask: URLSessionWebSocketTask?
    private var pingTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?
    private var reconnectTask: Task<Void, Never>?
    private var running = false
    private var isBackgroundMode = false
    private var reconnectAttempt = 0
    private var subscribedUserIds: Set<String> = []

    private let eventsContinuation: AsyncStream<Event>.Continuation
    public nonisolated let events: AsyncStream<Event>

    private let locationsContinuation: AsyncStream<LocationPoint>.Continuation
    public nonisolated let incomingLocations: AsyncStream<LocationPoint>

    public init(
        configuration: Configuration,
        session: URLSession = .shared
    ) {
        self.config = configuration
        self.session = session

        let (evStream, evCont) = AsyncStream<Event>.makeStream(bufferingPolicy: .bufferingNewest(256))
        self.events = evStream
        self.eventsContinuation = evCont

        let (locStream, locCont) = AsyncStream<LocationPoint>.makeStream(bufferingPolicy: .bufferingNewest(256))
        self.incomingLocations = locStream
        self.locationsContinuation = locCont
    }

    public func connect() async {
        running = true
        reconnectAttempt = 0
        await doConnect()
    }

    public func disconnect() async {
        running = false
        reconnectTask?.cancel()
        pingTask?.cancel()
        receiveTask?.cancel()
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        eventsContinuation.yield(.disconnected)
    }

    public func pushLocation(_ point: LocationPoint) async throws {
        let msg: [String: Any] = [
            "type": "location.push",
            "reqId": UUID().uuidString,
            "point": try encodedPoint(point)
        ]
        try await sendJSON(msg)
    }

    public func pushBatch(_ points: [LocationPoint]) async throws {
        let msg: [String: Any] = [
            "type": "location.batch",
            "reqId": UUID().uuidString,
            "points": try points.map { try encodedPoint($0) }
        ]
        try await sendJSON(msg)
    }

    public func subscribe(userId: String) async {
        subscribedUserIds.insert(userId)
        let msg: [String: Any] = ["type": "subscribe", "userId": userId]
        try? await sendJSON(msg)
    }

    public func unsubscribe(userId: String) async {
        subscribedUserIds.remove(userId)
        let msg: [String: Any] = ["type": "unsubscribe", "userId": userId]
        try? await sendJSON(msg)
    }

    /// Switch between foreground/background ping intervals.
    /// In background mode, ping interval increases to reduce radio wake-ups.
    /// If `disconnectInBackground` is enabled, the WebSocket is fully disconnected.
    public func setBackgroundMode(_ enabled: Bool) async {
        guard isBackgroundMode != enabled else { return }
        isBackgroundMode = enabled

        if enabled && config.disconnectInBackground {
            await disconnect()
            return
        }

        if !enabled && config.disconnectInBackground && !running {
            await connect()
            return
        }

        // Restart ping loop with appropriate interval
        if running {
            startPingLoop()
        }
    }

    // MARK: - Private

    private func doConnect() async {
        do {
            let token = try await config.tokenProvider.accessToken()
            let wsURL = makeWebSocketURL(from: config.baseURL)
            var request = URLRequest(url: wsURL)
            if !token.isEmpty {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }

            let ws = session.webSocketTask(with: request)
            self.webSocketTask = ws
            ws.resume()

            let authMsg: [String: Any] = ["type": "auth", "token": token]
            try await sendJSON(authMsg)

            reconnectAttempt = 0
            eventsContinuation.yield(.connected)

            for uid in subscribedUserIds {
                let subMsg: [String: Any] = ["type": "subscribe", "userId": uid]
                try? await sendJSON(subMsg)
            }

            startReceiveLoop()
            startPingLoop()
        } catch {
            eventsContinuation.yield(.error(error.localizedDescription))
            if running && config.autoReconnect {
                scheduleReconnect()
            }
        }
    }

    private func startReceiveLoop() {
        receiveTask?.cancel()
        receiveTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let ws = await self?.webSocketTask else { break }
                do {
                    let message = try await ws.receive()
                    let data: Data
                    switch message {
                    case .data(let d): data = d
                    case .string(let s): data = Data(s.utf8)
                    @unknown default: continue
                    }
                    await self?.handleServerMessage(data)
                } catch {
                    if await (self?.running ?? false) {
                        await self?.eventsContinuation.yield(.error(error.localizedDescription))
                        await self?.eventsContinuation.yield(.disconnected)
                        if await (self?.config.autoReconnect ?? false) {
                            await self?.scheduleReconnect()
                        }
                    }
                    break
                }
            }
        }
    }

    private func startPingLoop() {
        pingTask?.cancel()
        pingTask = Task { [weak self] in
            while !Task.isCancelled {
                let interval = await self?.activePingInterval ?? 30.0
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
                let msg: [String: Any] = ["type": "ping"]
                try? await self?.sendJSON(msg)
            }
        }
    }

    private var activePingInterval: TimeInterval {
        isBackgroundMode ? config.backgroundPingInterval : config.pingInterval
    }

    private func handleServerMessage(_ data: Data) {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        switch type {
        case "auth.ok":
            break
        case "location.ack":
            let reqId = json["reqId"] as? String ?? ""
            let status = json["status"] as? String ?? "accepted"
            eventsContinuation.yield(.pushAcknowledged(reqId: reqId, status: status))
        case "location.batch_ack":
            let reqId = json["reqId"] as? String ?? ""
            let ids = json["acceptedIds"] as? [String] ?? []
            eventsContinuation.yield(.batchAcknowledged(reqId: reqId, acceptedIds: ids))
        case "location.update", "location":
            if let pointData = json["point"],
               let pointJSON = try? JSONSerialization.data(withJSONObject: pointData),
               let point = try? RTLSJSON.decoder().decode(LocationPoint.self, from: pointJSON) {
                locationsContinuation.yield(point)
                eventsContinuation.yield(.locationReceived(point))
            }
        case "subscribed":
            eventsContinuation.yield(.subscribed(userId: json["userId"] as? String ?? ""))
        case "pong":
            break
        case "error":
            eventsContinuation.yield(.error(json["message"] as? String ?? "Server error"))
        default:
            break
        }
    }

    private func scheduleReconnect() {
        reconnectTask?.cancel()
        reconnectTask = Task { [weak self] in
            guard let self else { return }
            let attempt = await self.incrementReconnectAttempt()
            let exp = pow(2.0, Double(min(attempt - 1, 10)))
            let delay = min(await self.config.reconnectMaxDelay, await self.config.reconnectBaseDelay * exp)
            await self.eventsContinuation.yield(.reconnecting(attempt: attempt, delay: delay))
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            if await self.running { await self.doConnect() }
        }
    }

    private func incrementReconnectAttempt() -> Int {
        reconnectAttempt += 1
        return reconnectAttempt
    }

    private func sendJSON(_ obj: [String: Any]) async throws {
        let data = try JSONSerialization.data(withJSONObject: obj)
        let string = String(data: data, encoding: .utf8) ?? ""
        try await webSocketTask?.send(.string(string))
    }

    private func encodedPoint(_ point: LocationPoint) throws -> [String: Any] {
        let data = try RTLSJSON.encoder().encode(point)
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "RTLSWebSocket", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to encode point"])
        }
        return dict
    }
}

private func makeWebSocketURL(from baseURL: URL) -> URL {
    let httpURL = baseURL
        .appendingPathComponent("v1")
        .appendingPathComponent("ws")
    guard var components = URLComponents(url: httpURL, resolvingAgainstBaseURL: false) else {
        return httpURL
    }
    if components.scheme == "https" { components.scheme = "wss" }
    if components.scheme == "http" { components.scheme = "ws" }
    return components.url ?? httpURL
}
